### ISSUE TITLE
Feature request: Detecting game freezes in addition to crashes

### ISSUE THREAD

**Issue by TheDGOfficial** [Polyfrost/CrashPatch#10]
Similar to Spigot and Paper Watchdog, I would like if this mod could also catch game freezes on the main Client thread (preferably after 60 seconds of no ticking/rendering happening on the Client thread, a.k.a runGameLoop method being stuck somewhere). I am not sure how to display a menu with stack trace on the screen if main thread is frozen so perhaps it should write it to a file and exit the game (this what Spigot does it logs and exits) or use some desktop API to show the screen on another window.

While the main goal is to catch hard freezes (infinite loops, recursion, etc.) that don't recover (more than 60 seconds) it could also catch 5-10 second freezes like Paper with a config option to display stack trace in the log or a menu. In this case however it shouldn't exit the game since it's not a hard unrecoverable freeze, and instead of triggering while the freeze is ongoing, it should trigger after freeze has gone away (so that if it doesn't go away then the hard freeze of 60s will trigger instead).

<details>
  <summary>Spigot and Paper source</summary>

The Spigot's watchdog patch file can be found here: https://hub.spigotmc.org/stash/projects/SPIGOT/repos/spigot/browse/CraftBukkit-Patches/0023-Watchdog-Thread.patch
The Paper watchdog patch file can be found here (this patch file is applied on top of Spigot's watchdog java file): https://github.com/PaperMC/Paper/blob/master/patches/server/0227-Add-Early-Warning-Feature-to-WatchDog.patch

To get a Java file instead of a patch file with changes from both patches, build Paper locally and find the java file in sources folder.
</details>

<details>
  <summary>Explanation of how it works</summary>

Essentially what they do is create a thread in background with a long field set to System.currentTimeMillis() and update it every tick, and every second (10 seconds in Spigot) on the background thread compare this long field to the freshly-obtained System.currentTimeMillis(), and the difference between them should be the time since last tick has been run (or is still running). If this time is longer than 5 seconds Paper triggers soft freeze and if more than 60 seconds Spigot triggers hard freeze and exit with the thread dump of all threads (however most of the time only the Server / Client threads are relevant)
</details>

---

This mostly useful for finding freezes in mod code as vanilla code to my knowledge doesn't have any hard freezes and is lightweight with freezes except Garbage Collector. In the situation that a mod hard freezes you currently have no way of finding the source without using external tools like jstack which end users don't have access to or know using, and you have to kill the game from task manager which sometimes even affects OneConfig and resets config files of some mods if you happen to kill the game while a save is in progress.

<details>
  <summary>Details on the aforementioned config reset issue</summary>

Happened to me a few times, to my knowledge OneConfig triggers save of all configs about every 20 seconds even if  nothing changed on the configs and this might be the cause - this can be a different bug report but I'm not sure if its intented). So perhaps detecting a hard freeze inside MC and cleanly (not so clean since the game is frozen, but as close as possible to clean we can get, certainly cleaner then killing from task manager, at least) exiting (perhaps after ensuring no save in progress with a OneConfig hook) via System.exit would be better?

Actually probably no config hook is necessary since System.exit runs all shutdown hooks registered unlike killing from task manager which runs only some of them before they are being killed, see JavaDoc here: https://docs.oracle.com/javase/8/docs/api/java/lang/Runtime.html#addShutdownHook-java.lang.Thread- and https://docs.oracle.com/javase/8/docs/api/java/lang/Runtime.html#exit-int-

To my knowledge both Vigilance and OneConfig register a shutdown hook that flushes all config saves so if using System.exit likely no hook to check if a save is in progress is necessary and this might be the reason configs corrupt in the first place if the game is killed, quote from the JavaDoc:

"Shutdown hooks should also finish their work quickly. When a program invokes [exit](https://docs.oracle.com/javase/8/docs/api/java/lang/Runtime.html#exit-int-) the expectation is that the virtual machine will promptly shut down and exit. **When the virtual machine is terminated due to user logoff or system shutdown the underlying operating system may only allow a fixed amount of time in which to shut down and exit.**".

It does say user log off or system shutdown as an example but I'd guess killing from task manager also gives the shutdown hooks limited time to run before being interrupted and that results in corrupted config files.
</details>

**Comment by Wyvest** [comment]
Hey,

Thank you for the feature request. It seems like there has been a lot of effort put into this, and personally I think this is a pretty good idea!

However, we don't usually track feature requests here, so it could get lost within our already large suggestions list in our Discord. Could you make a suggestion in the suggestions channel at https://polyfrost.org/discord ?

Thanks!

**Comment by TheDGOfficial** [comment]
Sure thing, I have made one at https://discord.com/channels/822066990423605249/1276673372050427924

Should I close this issue then if its being tracked on Discord? Discord thread has link to the issue already so the issue won't be lost. I was not able to fit full description to discord's message limit and some of the markdown I used wasn't supported by Discord so I figured it'd be better to read it in GitHub so just placed a link to this issue in the thread description, hope that's fine.
